homeassistant/components/google_calendar.py

```diff
@@ -0,0 +1,407 @@
+        'https' if hass.http.use_ssl else 'http',
+        config.get(CONF_FQDN, get_fqdn()),
+        hass.http.server_address[1])
+
+    global _OAUTH_PATH_SETUP
+    if _OAUTH_PATH_SETUP:
+        return
+
+    from homeassistant.const import HTTP_OK
+    from oauth2client import client
+    flow = client.flow_from_clientsecrets(
+        hass.config.path(CONFIG_FILE),
+        scope=SCOPES,
+        redirect_uri="{}{}".format(base_url, CAL_SEARCH_CALLBACK)
+    )
+    # get a long term token with refresh_token
+    flow.params['access_type'] = 'offline'
+    flow.params['approval_prompt'] = 'force'
+
+    def _start_auth(handler, path_match, data):
+        """URL for starting Oauth session setup."""
+        # pylint disable=unused-argument
+        uri = flow.step1_get_authorize_url()
+        handler.send_response(301)
+        handler.send_header("Location", uri)
+        handler.end_headers()
+
+    def _finish_auth(handler, path_match, data):
+        """Oauth callback one authenticated with Google."""
+        if data.get("code") is None:
+            _LOGGER.error('Unknown error when authing')
+            response_message = """Something went wrong when attempting to
+                                authenticate with Google!"""
+        else:
+            from oauth2client.file import Storage
+            credentials = flow.step2_exchange(data.get("code"))
+            storage = Storage(hass.config.path(TOKEN_FILE))
+            storage.put(credentials)
+            do_setup(hass, config)
+            _LOGGER.info('Got credentials setup from Google')
+            response_message = """We're all setup now"""
+
+        html_response = """<html><head><title>Google Calendar Search</title>
+                </head><body><h1>{}</h1>
+                You can now close this window and restart Home-Assistant</br>
+                Restarting is not essential it will just cleanup the
+                </body>
+                </html>""".format(response_message)
+        html_response = html_response.encode("utf-8")
+        handler.send_response(HTTP_OK)
+        handler.write_content(html_response, content_type="text/html")
+
+    hass.http.register_path("GET",
+                            CAL_SEARCH_START,
+                            _start_auth,
+                            require_auth=False)
+    hass.http.register_path("GET",
+                            CAL_SEARCH_CALLBACK,
+                            _finish_auth,
+                            require_auth=False)
+
+    _OAUTH_PATH_SETUP = True
+    return
+
+
+def request_oauth_setup(hass, config):
+    """Request completion of oauth setup."""
+    from homeassistant.loader import get_component
+    configurator = get_component('configurator')
+    if DOMAIN in _CONFIGURING:
+        configurator.notify_errors(_CONFIGURING[DOMAIN],
+                                   "Failed to register, please try again.")
+        return False
+
+    from homeassistant.components.weblink import Link
```

Comment:
One component is not allowed to load another component to proxy functionality. We should add a link attribute to the configurator to start the auth flow.